File: panel/adb/packages.py
#!/usr/bin/env python3
"""Finding an application on the devices by a word out of its name.

The operator does not know the bundle id. They know the application by the
name on the display — "gebze", "kapi", the customer's own word — and typing
``com.piton.something`` correctly from memory is not a thing anybody does
twice. So the screen asks for a word and this module finds what matches.

The filtering happens HERE, not on the device. `pm list packages <word>` does
take a filter, but a locked-down Android build is not guaranteed to have the
shell utilities a pipeline would need and the filter's matching rules are the
package manager's rather than ours. Reading the whole list back and searching
it in Python costs one command and behaves identically on every device.

WHAT MAKES THIS MORE THAN A SEARCH BOX is that it runs on SEVERAL devices and
they need not agree. Four displays on a bench are routinely at different
stages of commissioning, and the honest answer to "which package matches
'gebze'?" is sometimes "this one, but only on three of them". That is exactly
the answer `search` gives: every match carries the devices it was found on
and the devices it is missing from. Collapsing that into a single list — the
obvious implementation — would have the operator press "start" on four
devices and be told nothing about the one that has no such application.
"""
from __future__ import annotations

import re
import threading
from concurrent.futures import ThreadPoolExecutor

from .. import i18n, settings
from ..errors import UnreachableError, user_message
from . import client

# `pm list packages` writes one per line, prefixed. Anything else on the line
# (a `-f` path, an installer name) is not asked for and not parsed.
_LINE = re.compile(r"^package:(\S+)", re.MULTILINE)
# A word short enough to match half the system matches the whole system; the
# operator gets 300 rows and no way to choose between them.
MIN_KEYWORD = 2


def parse_list(raw: str) -> list[str]:
    """Package names out of `pm list packages` output."""
    return sorted({name for name in _LINE.findall(raw or "")})


def installed(ip: str) -> list[str]:
    """Every package on one device.

    Raises rather than returning an empty list when the device cannot be
    reached: "no packages" and "no device" look identical on screen and mean
    opposite things.
    """
    if not client.connect(ip, attempts=2):
        raise UnreachableError(i18n.t("error.adbNoConnection"))
    return parse_list(client.shell(ip, "pm", "list", "packages"))


def keywords(raw) -> list[str]:
    """The words to search for, out of one comma-separated box.

    SEVERAL AT ONCE, because a bench rarely holds one kind of display. Four
    devices running three different customers' applications is the ordinary
    case, and searching them one word at a time means three searches, three
    selections and three separate runs — while the operator wanted one.

    Semicolons are accepted as well: a list pasted out of a spreadsheet
    arrives with whichever separator that spreadsheet used.
    """
    words = [word.strip()
             for word in str(raw or "").replace(";", ",").split(",")]
    return [word for word in dict.fromkeys(words) if word]


def matches(names, words) -> list[str]:
    """Names containing ANY of these words."""
    wanted = [str(word).strip().lower() for word in words if str(word).strip()]
    return [name for name in names
            if any(word in name.lower() for word in wanted)]


def search(ips, keyword) -> dict:
    """Which packages match, and on which of these devices.

    The reply is shaped for the screen:

        {"keywords": […],
         "packages": [{"name": …, "present": [ip…], "missing": [ip…]}],
         "failed":   [{"ip": …, "error": …}]}

    A device that could not be read is in `failed` and NOT in any package's
    `missing`: "the application is not installed here" and "this address did
    not answer" are two different problems with two different next steps.

    `present` is also what the operation bar builds its targets from, so a
    device that does not carry a package is never sent a command about it.
    """
    words = keywords(keyword)
    if not words:
        raise ValueError(i18n.t("error.adbKeywordTooShort", count=MIN_KEYWORD))
    short = [word for word in words if len(word) < MIN_KEYWORD]
    if short:
        # One short word in a list of good ones would otherwise pull the
        # whole system in beside the three packages that were wanted.
        raise ValueError(i18n.t("error.adbKeywordTooShort", count=MIN_KEYWORD))

    targets = [str(ip) for ip in ips if str(ip or "").strip()]
    if not targets:
        raise ValueError(i18n.t("error.adbNoDeviceSelected"))

    found: dict[str, list[str]] = {}
    failed: list[dict] = []
    reached: list[str] = []
    # The devices are read in parallel and all three collections above are
    # written from those threads.
    lock = threading.Lock()

    def one(ip: str) -> None:
        try:
            names = matches(installed(ip), words)
        except Exception as exc:
            with lock:
                failed.append({"ip": ip, "error": user_message(exc)})
            return
        finally:
            client.disconnect(ip)
        with lock:
            reached.append(ip)
            for name in names:
                found.setdefault(name, []).append(ip)

    pool = ThreadPoolExecutor(
        max_workers=min(settings.ADB_WORKERS, len(targets)))
    try:
        list(pool.map(one, targets))
    finally:
        pool.shutdown(wait=True)

    packages = [{
        "name": name,
        "present": sorted(present),
        # Only devices that answered. See the docstring.
        "missing": sorted(ip for ip in reached if ip not in set(present)),
    } for name, present in sorted(found.items())]
    return {"keywords": words, "packages": packages,
            "failed": sorted(failed, key=lambda row: row["ip"])}

File: static/js/views/adb/packages.js
// Finding the applications by a word or two, and saying honestly what was
// found.
//
// The operator does not know the bundle id and should not have to. They type
// a word out of the name — "gebze" — and the devices are asked which of their
// packages contain it. SEVERAL WORDS, comma separated, because a bench rarely
// holds one kind of display: four devices running three customers'
// applications is the ordinary case, and one search that finds all three
// beats three searches.
//
// SEVERAL ANSWERS CAN BE CHOSEN, for the same reason. The operation bar then
// works on (device, bundle) pairs — see state.operationTargets — so
// "restart the application on all of them" means a different package name on
// each device and still one run.
//
// WHAT THIS SCREEN REFUSES TO HIDE. Several displays are searched at once and
// they need not agree: three have the application and the fourth does not,
// because somebody has not installed it yet. The obvious presentation — one
// list of package names — makes that fourth display invisible, and the
// operator finds out about it by pressing "start" and watching one row fail
// for no visible reason. So every result carries the devices it was found on,
// and the ones it is missing from are named.
//
// A device that could not be READ is a third thing again and is listed
// separately: "the application is not installed here" and "this address did
// not answer" have different fixes.

import { el } from '../../core/dom.js';
import { t } from '../../core/i18n.js';
import { local, running, selectedIps } from './state.js';
import { tagged } from './fields.js';

export function packagesCard(actions) {
  const answer = local.found;
  const chosen = selectedIps();

  const keywordField = el('input', {
    class: 'field adb-keyword', type: 'text', autocomplete: 'off',
    spellcheck: 'false', value: local.keyword,
    oninput: (event) => { local.keyword = event.target.value; },
  });

  return el('section', { class: 'card corner adb-packages' }, [
    el('div', { class: 'card-head' }, [
      el('h3', { text: t('adb.application') }),
    ]),
    el('form', {
      class: 'adb-search',
      onsubmit: (event) => {
        event.preventDefault();
        // Focus has to leave the box or the results never draw: a focused
        // field holds the render back (see app.focusInScreenField).
        keywordField.blur();
        actions.search(keywordField.value);
      },
    }, [
      tagged('adb.keyword', keywordField),
      el('button', {
        type: 'submit', class: 'btn btn-primary',
        text: local.searching ? t('adb.searching') : t('adb.search'),
        disabled: local.searching || !chosen.length || running(),
        title: chosen.length ? '' : t('adb.selectDeviceFirst'),
      }),
    ]),
    ...results(answer, actions),
  ]);
}

function results(answer, actions) {
  if (!answer) return [];
  const found = answer.packages || [];
  const failed = answer.failed || [];
  const parts = [];

  if (!found.length) {
    parts.push(el('p', {
      class: 'info',
      text: t('adb.noPackageMatched',
              { word: (answer.keywords || []).join(', ') }),
    }));
  } else {
    parts.push(el('div', { class: 'adb-package-list' },
                 found.map(entry => packageRow(entry, actions))));
  }

  for (const row of failed) {
    parts.push(el('p', {
      class: 'warning',
      text: t('adb.deviceNotRead', { ip: row.ip, error: row.error }),
    }));
  }
  return parts;
}

// THE ADDRESSES IT IS ON, not the ones it is not. Naming the gaps was the
// first attempt and it reads backwards: the operator is choosing what to work
// on, so the useful list is the one the run will actually reach. When some
// devices are missing it, the line still says so — as a count, which is the
// part that matters — and the row keeps its warning colour.
function packageRow(entry, actions) {
  const on = local.packages.has(entry.name);
  const present = entry.present || [];
  const missing = entry.missing || [];
  return el('button', {
    type: 'button', class: 'checkbox adb-package',
    'aria-pressed': String(on),
    onclick: () => actions.choosePackage(entry.name),
  }, [
    el('span', { class: 'box', 'aria-hidden': 'true' }),
    el('span', { class: 'adb-package-text' }, [
      el('span', { class: 'mono truncate', text: entry.name }),
      el('span', {
        class: missing.length ? 'adb-package-warn' : 'adb-package-note',
        text: missing.length
          ? t('adb.presentOnSomeDevices', {
            list: present.join(', '), missing: missing.length,
          })
          : t('adb.presentOnAllDevices', { list: present.join(', ') }),
      }),
    ]),
  ]);
}
